### DISCUSSION TITLE
tutorials/lift-off-part1/01-feature-overview-and-setup

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#37] [course feedback]
# tutorials/lift-off-part1/01-feature-overview-and-setup

https://www.apollographql.com/tutorials/lift-off-part1/01-feature-overview-and-setup

**Comment by CMH7** [comment]
Such a straightforward tutorial, I love it <3

**Reply by Jonathon** [reply]
So far so good, easy to follow. 9/10

**Comment by Theedon** [comment]
nice

**Comment by jgerton** [comment]
alright! looking forward to some graphql tutorials!  @jgerton gettin' it done! #100DaysOfCode

**Reply by CMH7** [reply]
I would like to do that to that #100DaysOfCode

**Comment by yunushabeeb** [comment]
A straight forward tutorial. No excessive analysis. Getting to love APOLLO...

**Comment by adetorodev** [comment]
Doing the tutorial again. Nice one!

**Comment by Marzooq13579** [comment]
A clean walkthrough. Love it!

**Comment by alperozhan** [comment]
loved the kittycats, here we go!

**Comment by mathoks** [comment]
Awesome it all checked out

**Comment by attilacsanyi** [comment]
So far so good, excited to continue 👍

**Comment by marcusj89** [comment]
Hard work and good wil.  wu hoo

**Reply by marcusj89** [reply]
meant goodwill like the goodwill stores

**Comment by fcholak** [comment]
Awesome training material! Can't wait to get through it.

**Comment by dasUjjwal1** [comment]
Nice

**Comment by lovephimu** [comment]
Nice

**Comment by anvodev** [comment]
Nice!

**Comment by adriatic** [comment]
All comments are and should be positive as this is a terrific start of. very prospective tutorial. Great work.

**Comment by utkarshraj1** [comment]
Client is not starting as showing some loader error.

**Reply by lizhennessy** [reply]
Hi there @utkarshraj1, I just saw your comment in the Odyssey forum as well. You should have a file called `vite.config.js` in the root of your project directory; this file tells Vite to interpret all of our `.js` files as `.jsx`, and it appears it might be responsible for this error you're seeing. Could you check to see if that file is present? Let me know and I'll be happy to troubleshoot further!

**Reply by utkarshraj1** [reply]
Hi,
Thanks for reaching out. The vite.config file is present and the loader is also set as "jsx". Still it is showing the jsx js error

**Reply by lizhennessy** [reply]
Hey there - hm, that is interesting. Could you let me know what version of Node you're running? (By running `node --version` in your terminal?)

**Reply by utkarshraj1** [reply]
The node version is 16.19.1

**Reply by lizhennessy** [reply]
Thanks! Looks like that's probably not our issue. Do you have a repository pushed to GitHub that I could take a look at?

**Reply by utkarshraj1** [reply]
Yes, sure. Please find the link below:

https://github.com/utkarshraj1/apollo-certification-study

**Reply by lizhennessy** [reply]
Thanks so much for your patience, @utkarshraj1. Are you working on a Windows computer, by chance? It seems like we might have some incompatibility issues with the code in its current state. 

If so, I made an updated branch that I think might get us around the error. You can reclone it by running the following command, or by checking out the `jsx` branch from the parent repository - effectively I just changed all of the `.js` file extensions containing JSX code to `.jsx` files as it seems the loader is giving us a bit of trouble. In the meantime, I am going to be troubleshooting Vite to see if there's anything that I've missed to enable its functionality on Windows in addition to Mac. Please let me know if this works for you, or if you continue to run into any trouble!

```
git clone -b jsx https://github.com/apollographql/odyssey-lift-off-part1
```

**Reply by lizhennessy** [reply]
@utkarshraj1 I think I've found a solution. In your `vite.config.js` file, try adding the following lines to your `esbuildOptions` object, just above the `plugins` key. This should solve the compatibility issue with Windows!

```
loader: {
        '.js': 'jsx',
      },
```

**Reply by utkarshraj1** [reply]
Thanks, the issue is resolved and I am able to run the client on my Windows machine. Not sure why it wasn't working earlier despite making the same changes. I guess I need to understand more about Vite!

**Reply by lizhennessy** [reply]
So happy to hear that, @utkarshraj1! Thank you so much for raising the issue and helping us improve the course for Windows users.

**Comment by AjayBains** [comment]
Getting error while trying to run client 
error when starting dev server:
Error: getaddrinfo ENOTFOUND localhost
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:internal/dns/promises:42:17)

**Reply by AjayBains** [reply]
oh,this seems to b my system issue, facing same with few other projects too

**Reply by AjayBains** [reply]
So its happening in my local but only for projects stated with vite

**Reply by lizhennessy** [reply]
Thanks for letting us know, @AjayBains! I'm curious whether the `localhost` alias might not be set up in your computer's `/etc/hosts` file; are you able to open up that file and see if it exists? You might need it to contain a line similar to the below.

```
127.0.0.1	localhost
```

Another option is to open up the `vite.config.js` file in the `client` directory and change the `server` key in the config option to point to a different host, such as what's shown below:

```
 server: {
    host: '127.0.0.1',
    port: 3000
  },
```

**Reply by AjayBains** [reply]
updating vite.config worked,thank you so much

**Reply by lizhennessy** [reply]
Excellent! So happy to hear it. Please reach out if anything else pops up!

**Comment by wesley974** [comment]
Awesome! 

Can someone explain the role of the packages located in the client? 
I have never used the following ones:
```
    "@apollo/space-kit": "^9.3.1",
    "@emotion/cache": "^11.4.0",
    "@emotion/core": "^10.1.1",
    "@emotion/react": "^11.4.0",
    "@emotion/styled": "^11.3.0",
    "framer-motion": "^4.1.17",
    "react-emotion": "^10.0.0",
    "react-markdown": "^6.0.2",
    "react-player": "^2.6.0",
```

**Reply by lizhennessy** [reply]
Hi there @wesley974 ! Happy to clarify. 

The `@emotion/` libraries are all there to help us with the styling of components. The `@emotion/react` package gives us React-specific tooling right out of the box, and with `@emotion/styled` we can define styled React components right in our JSX files. Additionally, the `@apollo/space-kit` package is Apollo's style library, which gives us some of our fundamental CSS rules, components and colors - just to speed up some of the styling that we include in the Catstronauts app!

`react-player` gives us the video utility that we'll use in `Lift-off Lab` to render and play videos as part of our different `Track`s! The `react-markdown` package gives us the ability to write and render text in Markdown format within our app. Some of the data we'll learn how to query and consume in the course of the Lift-off series returns content written in Markdown, which is where this package will come in handy. These two packages are here as part of the boilerplate so that learners don't have to worry about adding them into the project later - they can focus on the GraphQL and Apollo concepts instead!

As far as `react-emotion` and `framer-motion` go, these might be unnecessary dependencies - I'll do a bit of further research to see if we're using them anywhere in the code, and if not I'll make sure they're removed from the starter code. Thanks for bringing it to our attention!

**Comment by MissyM** [comment]
Thanks!

**Comment by Sharonah8** [comment]
Getting this error: "UNABLE_TO_GET_ISSUER_CERT_LOCALLY" when trying to run npm install

**Comment by brencon** [comment]
Very straightforward and easy to follow. Thank you!

**Comment by amt-dwemoh-kesseh** [comment]
Straight to the point

**Comment by Rudrasish** [comment]
Simple and straightforward for any beginner.

**Comment by saiyantanghosh** [comment]
working

**Comment by doanwebute** [comment]
very goood tutorial, thank you so much

**Comment by EjDadivas** [comment]
I like how you scaffold the topics and unpack the ideas! Great material

**Comment by Stark-Akshay** [comment]
Really cool content, great for learning.

**Comment by repiemm** [comment]
Up and running in a few minutes, pretty straightforward!

**Comment by Femtech-web** [comment]
The most easy and straightforward tutorial i have ever taken.
kudos!!

**Comment by le-el** [comment]
cool course

**Comment by nonopolarity** [comment]
I like the tutorial, which leads you to know about every aspect of the project. In some companies I work for, it is like, "Ok, this is day 1 since you joined us. Now we won't give you any docs, wiki, but just add the front end feature 1 and backend feature 1 to our project. Oh yeah, we won't even tell you what the project is about and how to use the app. Just add features to it. When you are done, I will come take your credit."

**Comment by lycha0206** [comment]
Is it true that this comment feature from GitHub doesn't allow editing or removal of the comment once it is here? Then we'd better double check the content before posting. Is it possible to at least edit it?

**Reply by lizhennessy** [reply]
Hi there @lycha0206 - I can take care of deleting comments if you've posted one and would prefer to remove it. Which comment would you like to remove?

**Reply by lycha0206** [reply]
not at the moment. Just that sometimes I type a comment and want to edit it a little, such as if I forgot to close the `"` quote or something, I hope to be able to do it myself. Maybe Github doesn't want this edit feature because it can lead to all kinds of issues if 3 or 5 people replied and the original author edits the comment to something else.

**Comment by belloshehu** [comment]
Short and clear! No long story. I love it!

**Comment by LingVuDev** [comment]
Such a beautiful tutorial

**Comment by NishantCoder108** [comment]
Feature overview and setup ,Done 🔥

**Comment by vishwaefor** [comment]
Is Vite hot reload working?

**Reply by NishantCoder108** [reply]
Yes, it is working , it's the feature of vite, Hot reloading.

**Comment by azimaro** [comment]
I installed dependencies in the server directory, but when I try to run the applicable command in the client dir I am getting a lot of deprecation errors:
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated abab@2.0.6: Use your platform's native atob() and btoa() methods instead
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated domexception@2.0.1: Use your platform's native DOMException instead
npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
npm ERR! code EEXIST
npm ERR! syscall mkdir
npm ERR! path /Users/angelinazimaro/.npm/_cacache/content-v2/sha512/45/1d
npm ERR! errno -13
npm ERR! EACCES: permission denied, mkdir '/Users/angelinazimaro/.npm/_cacache/content-v2/sha512/45/1d'
npm ERR! File exists: /Users/angelinazimaro/.npm/_cacache/content-v2/sha512/45/1d
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.

What am I doing wrong?

**Comment by le-el** [comment]
@azimaro 
The errors you're encountering are primarily due to deprecated packages and a permission issue on your system. Let's address these concerns step by step:

 - Deprecated Packages: The warnings about deprecated packages (source-map-url, urix, abab, resolve-url, source-map-resolve, domexception, w3c-hr-time) indicate that these packages are no longer maintained or recommended for use. For each deprecated package, visiting the provided GitHub links will offer insights into why they're deprecated and suggest alternatives or updates. For example, resolve-url and urix are deprecated, and their repositories might provide guidance on updated practices or alternative packages.
Permission Denied Error: The EACCES: permission denied error suggests a permissions issue with the .npm directory in your user folder. This can happen if npm was run with sudo or as a root user at some point, causing some files or directories to be owned by the root user. To resolve this, you can take ownership of the .npm directory with the following command:
sudo chown -R $(whoami) ~/.npm
This command changes the ownership of the .npm directory to the current user, resolving the permission denied error.
EEXIST Error: The EEXIST error indicates that npm is trying to create a directory that already exists. This could be a symptom of the permission issues mentioned above or a corrupted npm cache. After resolving the permissions issue, try clearing the npm cache with:
   ~   npm cache clean --force   ~
Note that using --force is generally not recommended unless necessary, as it bypasses some safety checks. However, in this case, it can help resolve cache-related issues.
After addressing these issues, try running your npm command again in the client directory. If you continue to encounter problems with deprecated packages, consider updating your dependencies to newer, maintained versions that provide similar functionality.

**Reply by azimaro** [reply]
Thank you! I will try these suggestions

**Reply by azimaro** [reply]
Changing owner of .npm seemed to work, but cache cleaning says permission denied, so maybe not.

**Reply by azimaro** [reply]
I ended up reinstalling node, seems to have fixed

**Comment by juileekharat** [comment]
Good so far!!

**Comment by gustvoassuncao** [comment]
Thank you for the tutorial! It was awesome!

**Comment by piyushyadav1617** [comment]
Amazing tutorial so far!

**Comment by lrojasv** [comment]
HI, I have a problem when i run client the command npm install && npm start
The error is 

Could not resolve "./cjs/react-is.development.js"

    node_modules/prop-types/node_modules/react-is/index.js:6:27:      
      6 │   module.exports = require('./cjs/react-is.development.js');
        ╵                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  

X [ERROR] Failed to resolve entry for package "extend". The package may have incorrect main/module/exports specified in its package.json. [plugin vite:dep-pre-bundle]

    node_modules/esbuild/lib/main.js:1373:21:
      1373 │         let result = await callback({
           ╵                      ^

**Reply by lizhennessy** [reply]
hi there @lrojasv ! Happy to take a look. What version of Node are you running? And are you running the TypeScript or the JavaScript version of the course?

**Comment by KoolieKangarooMjolner** [comment]
Working so far! :D
but project README.md states ```npm install && npm start``` while this page states ```npm install && npm run dev``` (when installing dependencies and starting server)

**Reply by mabuyo** [reply]
Thanks for letting us know about the discrepancy @KoolieKangarooMjolner. The course contents are the most up-to-date, but either of these start instructions should work!

**Comment by nsauce** [comment]
I noticed when installing and running npm for client and server, a double "&&" symbol is used for those commands. This works bash terminal, but if you have PowerShell terminal in your visual studio code, it will not like the double "&&" it like single form "&". Thought it would be worth mentioning. Seems this assumes you are in bash terminal.

**Reply by KoolieKangarooMjolner** [reply]
I would not recommend using PowerShell for node based projects - use bash or any bash derivate - even in a windows environment. I use ZSH (Z Shell) with [Oh My ZSH](https://ohmyz.sh/) which is a great experience! I remember installing ZSH with Oh My ZSH when I worked for a client that required me to use a PC. It was not as easy as on a MAC but it can be done. Think this was the guide I followed: https://dev.to/zinox9/installing-zsh-on-windows-37em (Sorry for keeping this off topic path alive) :D

**Reply by mabuyo** [reply]
Hey there @nsauce - thank you for pointing this out! It's now on our list to fix.

**Comment by SumeetWajpe** [comment]
Hey Team !

I would like suggest a amended in the command that allows you to run the server ! Its mentioned "npm install && npm run dev", but the package.json has script named "start" and not dev.

**Reply by mabuyo** [reply]
Hey there @SumeetWajpe ! Apologies for the late reply. To clarify, the TypeScript version of the course includes a `dev` command, whereas the JavaScript version only has `start`. Hope that helps!

**Comment by karloteke** [comment]
Nice

**Comment by RAD-AgileDevOps** [comment]
Simple and motivational tutorial

**Comment by huy312100** [comment]
Brillant tutorial. Very detail !

**Comment by albeertito7** [comment]
so far so goood

**Comment by illuminoir** [comment]
Amazing tutorial, as everyone mentioned, straight forward and very easy to navigate - thank you guys!

**Comment by rBo3Db** [comment]
hi guys. "npm run dev" for the server in ts tutorial - do you mean "npm run start" mb?

**Reply by mabuyo** [reply]
Hey there, the `npm run dev` command enables hot-reloading so you don't have to restart the server every time you make changes!

**Comment by SamuelLeiva** [comment]
So far so good
